### ISSUE TITLE
Is `v = np.array(v.dt.to_pydatetime())` still necessary?

### ISSUE THREAD

**Issue by MarcoGorelli** [plotly/plotly.py#4836]
As far as I can tell, the lines

https://github.com/plotly/plotly.py/blob/960adb9b9a89387d05343497de1df5d3df592698/packages/python/plotly/_plotly_utils/basevalidators.py#L101-L108

were introduced in https://github.com/plotly/plotly.py/pull/1163 to introduce issues with displaying numpy datetime64 arrays

However, have the numpy datetime64 issues since been fixed? From having built Polars from source, then here's what I see on the master branch:

![Image](https://github.com/user-attachments/assets/3314a83d-d495-4ef0-bf65-d551b6a80c9a)

Looks like it displays fine

If I apply the diff
```diff
--- a/packages/python/plotly/_plotly_utils/basevalidators.py
+++ b/packages/python/plotly/_plotly_utils/basevalidators.py
@@ -95,20 +95,7 @@ def copy_to_readonly_numpy_array(v, kind=None, force_numeric=False):
 
     # Handle pandas Series and Index objects
     if pd and isinstance(v, (pd.Series, pd.Index)):
-        if v.dtype.kind in numeric_kinds:
-            # Get the numeric numpy array so we use fast path below
-            v = v.values
-        elif v.dtype.kind == "M":
-            # Convert datetime Series/Index to numpy array of datetimes
-            if isinstance(v, pd.Series):
-                with warnings.catch_warnings():
-                    warnings.simplefilter("ignore", FutureWarning)
-                    # Series.dt.to_pydatetime will return Index[object]
-                    # https://github.com/pandas-dev/pandas/pull/52459
-                    v = np.array(v.dt.to_pydatetime())
-            else:
-                # DatetimeIndex
-                v = v.to_pydatetime()
+        v = v.values
```
then it looks like pandas datetime Series still display fine

![Image](https://github.com/user-attachments/assets/4a73a300-1e86-4875-b1fb-559e2db5def4)

---

Asking in the context of https://github.com/plotly/plotly.py/pull/4790, as `copy_to_readonly_numpy_array` would need to handle other kinds of inputs (not just pandas series / index)

A plain conversion to numpy would be a lot faster than going via stdlib datetime objects
```
In [23]: %time np.array(s.dt.to_pydatetime())
CPU times: user 325 ms, sys: 8.34 ms, total: 333 ms
Wall time: 360 ms
Out[23]:
array([datetime.datetime(2000, 1, 1, 0, 0),
       datetime.datetime(2000, 1, 1, 1, 0),
       datetime.datetime(2000, 1, 1, 2, 0), ...,
       datetime.datetime(2114, 1, 29, 13, 0),
       datetime.datetime(2114, 1, 29, 14, 0),
       datetime.datetime(2114, 1, 29, 15, 0)], dtype=object)

In [24]: %time s.to_numpy()
CPU times: user 46 μs, sys: 0 ns, total: 46 μs
Wall time: 51.5 μs
Out[24]:
array(['2000-01-01T00:00:00.000000000', '2000-01-01T01:00:00.000000000',
       '2000-01-01T02:00:00.000000000', ...,
       '2114-01-29T13:00:00.000000000', '2114-01-29T14:00:00.000000000',
       '2114-01-29T15:00:00.000000000'], dtype='datetime64[ns]')
```

**Comment by emilykl** [comment]
I did a bit more digging on this issue to gather more background — @MarcoGorelli @Fbruzzesi Hopefully this lines up with what you've found.

As described in #1160, the reported bug in `plotly==3.2.0` was that pandas datetime series passed into figures became transformed into integers — for example:
```python
df = pd.DataFrame({
    'date': pd.to_datetime(["2024-01-01 01:00", "2024-01-01 02:00"]),
    'value': [100, 200],
})
fig = go.Figure(
    data=[go.Scatter(x=df['date'], y=df['value'])],
)
print(fig.data[0].x.__repr__())
```
produced the output
```python
array([1704070800000000000, 1704074400000000000], dtype=object)
```
and resulted in the following chart:

<details>
  <summary>Show chart</summary>

![Image](https://github.com/user-attachments/assets/d550decc-781b-4eb3-a627-6694206a7318 )
</details>


---

This bug was a side effect of the changes in #1149 (specifically it seems the addition of `v = v.values` caused the lower-down line `new_v = np.array(v, dtype='object')` to convert an array of pandas datetimes to an array with `dtype=object` whose individual items are integers). 

So, #1163 revised the logic to call `to_pydatetime` (which returns an ndarray) on datetime series, rather than assigning `v = v.values`. After that change (first released in `plotly==3.2.1`), pandas datetime series were converted to ndarrays of python datetimes in `fig.data`:
```python
array([datetime.datetime(2024, 1, 1, 1, 0), datetime.datetime(2024, 1, 1, 2, 0)], dtype=object)
```
and the chart was plotted correctly:


<details>
  <summary>Show chart</summary>

![Image](https://github.com/user-attachments/assets/84840b0b-5163-45d7-b50d-4e70a00af99c)
</details>

(Incidentally, the behavior of `to_pydatetime()` returning an ndarray [is deprecated](https://pandas.pydata.org/docs/reference/api/pandas.Series.dt.to_pydatetime.html) and it returns a pandas Series as of `pandas==2.2`, so the behavior has already changed from what was intended.)

A [test](https://github.com/plotly/plotly.py/blob/6319ab9cd35a3dfd883c1a8f050af84b1a0039a1/packages/python/plotly/_plotly_utils/tests/validators/test_pandas_series_input.py#L176C1-L190C1) was also added in #1163 to verify that when passing a pandas datetime series as input, the corresponding `fig.data` key consists of python datetime objects. 

---

The latter behavior holds to this day on `master` — pandas datetime series passed into a figure become ndarrays of python datetime objects in `fig.data`.

_HOWEVER_, with the [current implementation of the narwhals adoption](https://github.com/plotly/plotly.py/pull/4790), pandas datetime series now become ndarrays of `np.datetime64` objects in `fig.data`.
```python
array(['2024-01-01T01:00:00.000000000', '2024-01-01T02:00:00.000000000'], dtype='datetime64[ns]')
```

The test added in #1163 was also modified to check that the `fig.data` key consists of `np.datetime64` rather than python datetimes.

So the real question is — is that a problem? It seems to me that as long as they serialize to the same JSON as python datetimes, and the resulting charts are identical, it shouldn't be an issue — though of course I may be missing something. Thanks to the detailed work of @MarcoGorelli and @Fbruzzesi, the JSON is... almost... identical:

on `master`:
```json
"x":["2024-01-01T01:00:00","2024-01-01T02:00:00"]
```

on Narwhals branch:
```json
"x":["2024-01-01T01:00:00.000000000","2024-01-01T02:00:00.000000000"]
```

The only difference I can see is the precision, I am not sure we need nanosecond precision but it also seems silly to remove it if it comes by default. The generated charts are identical.

**Comment by MarcoGorelli** [comment]
thanks all - closing as addressed by https://github.com/plotly/plotly.py/pull/4790
